Add SCSR parsing to RDKit

tadhurst-cdd opened the pull request:
<!--
Thanks for contributing a pull request! 
-->
#### Reference Issue
Support for SCSR  (Self-contained Sequence Representation) in RDKit


#### What does this implement/fix? Explain your changes.
Adds parsing of SCSR files into RDKit as new entity (SCSRMol) and converts SCSRMols to mols with S-groups.

#### Any other comments?

A supporting documents is:

[ScsrSupportInRDKit.pdf](https://github.com/user-attachments/files/18708315/ScsrSupportInRDKit.pdf)

One example file is:



[Mixed .txt](https://github.com/user-attachments/files/18286846/Mixed.txt)

whosayn commented on Code/GraphMol/FileParsers/MolFromScsr.cpp:
```diff
@@ -0,0 +1,494 @@
+//  of the RDKit source tree.
+//
+
+#include "MolSGroupParsing.h"
```
The windows build failure might be due to not including  `FileParsers.h` for the `MolFromScsr` definition

tadhurst-cdd replied:
MolFromScsr.cpp has been removed and the code that was in it is now in ScsrMolFileParser.cpp.   That file contains a reference to #include "FileParserUtils.h", which includes "FileParsers.h".

greglandrum reviewed (commented):
I took a look at the API and the data structures (i.e. only looking at header files and the cpp file for the python wrapper) and here's a set of questions, comments, and suggestions.

tadhurst-cdd commented on Code/GraphMol/FileParsers/FileParsers.h:
```diff
@@ -63,6 +69,19 @@ RDKIT_FILEPARSERS_EXPORT std::unique_ptr<RWMol> MolFromMolFile(
+    const MolFileParserParams &params = MolFileParserParams());
+RDKIT_FILEPARSERS_EXPORT std::unique_ptr<RWMol> MolFromScsr(
+    const RDKit::SCSRMol &scsrMol,
+    const MolFromScsrParams &molFromScsrParams = MolFromScsrParams());
```
ok

tadhurst-cdd commented on Code/GraphMol/FileParsers/FileParsers.h:
```diff
@@ -53,6 +53,12 @@ struct RDKIT_FILEPARSERS_EXPORT MolFileParserParams {
+
+struct RDKIT_FILEPARSERS_EXPORT MolFromScsrParams {
+  bool includeLeavingGroups =
+      true; /* when true, leaving groups on atoms that are not exo-bonded are
```
OK

tadhurst-cdd commented on Code/GraphMol/FileParsers/FileParsers.h:
```diff
@@ -63,6 +69,19 @@ RDKIT_FILEPARSERS_EXPORT std::unique_ptr<RWMol> MolFromMolFile(
     const std::string &fName,
     const MolFileParserParams &params = MolFileParserParams());
 
+RDKIT_FILEPARSERS_EXPORT std::unique_ptr<RDKit::SCSRMol> ScsrFromScsrDataStream(
```
this change makes it more consistem with ROMol and RWMol routines,  Changed

tadhurst-cdd commented on Code/GraphMol/FileParsers/MolSGroupParsing.h:
```diff
@@ -22,24 +22,24 @@ typedef std::map<int, STR_VECT> IDX_TO_STR_VECT_MAP;
 /* ------------------ V2000 Utils  ------------------ */
 
-unsigned int ParseSGroupIntField(const std::string &text, unsigned int line,
+unsigned int ParseSGroupIntField(const std::string &text, unsigned int &line,
```
the routine that started this was ParseV3000SGroupsBlock which updated the line value that was passed to it as a value.  This cause error reports that had wrong lines.  I changed all the calls to reference, even the routines that do not update line, just so the caller does not have to know if the routine updates the line count or not.

greglandrum replied:
> the routine that started this was ParseV3000SGroupsBlock which updated the line value that was passed to it as a value. This cause error reports that had wrong lines. I changed all the calls to reference, even the routines that do not update line, just so the caller does not have to know if the routine updates the line count or not.

The problem with that logic is that the caller *can't* know if the route updates the line or not (without reading the code). Passing the line number as a non-const reference is a signal that the line number will/may be changed. I think this is useful

tadhurst-cdd replied:
OK I did this.  But this all should really be in a class, and "line" should be a member that all methods share.    Making all routines that use the same reference to line sort of emulates that.   I thought your comment from above about "the caller not needing to know if line changed or not, and that this was useful" meant NOT to change them back.    So I just had to read each routine to see if it did or could change the value of line.

greglandrum replied:
> OK I did this. But this all should really be in a class, and "line" should be a member that all methods share.

I can see that. But in a separate PR please. ;-)

tadhurst-cdd commented on Code/GraphMol/Wrap/Mol.cpp:
```diff
@@ -828,6 +843,40 @@ struct mol_wrapper {
+
+    python::class_<RDKit::SCSRMol, SCSRMOL_SPTR, boost::noncopyable>(
+        "SCSRMol", SCSRmolClassDoc.c_str(),
+        python::init<>(python::args("self"), "Constructor, takes no arguments"))
```
This was copied from ROMol and/or RWMol.  I can think of no use for this from python 
.  I removed it

tadhurst-cdd commented on Code/GraphMol/Wrap/Mol.cpp:
```diff
@@ -828,6 +843,40 @@ struct mol_wrapper {
+
+        .def(python::init<const RDKit::SCSRMol &>(
+            (python::arg("self"), python::arg("scsr"))))
+        .def("__copy__", &generic__copy__<RDKit::SCSRMol>, python::args("self"))
```
THese were copies from ROMol.   I don't know what they are for.   I deleted them and the tests still run

tadhurst-cdd commented on Code/GraphMol/Atom.h:
```diff
@@ -37,6 +37,21 @@ namespace RDKit {
+class AtomAttchOrd;  // class for ATTAHORD of the form  ATTCHORD=(4 1 Al 3 Br)
+
+class RDKIT_GRAPHMOL_EXPORT AtomAttchOrd
+    : public std::pair<unsigned int, std::string> {
```
This went through iterations.  At one point, the first and second functions were called directly.   There is no need for that now that the getAtomIdx and getLabel functions are defined.  So I changed it as suggested to a private data member, which protects other parts of pair from access.

tadhurst-cdd commented on Code/GraphMol/PeriodicTable.h:
```diff
@@ -90,7 +90,7 @@ class RDKIT_GRAPHMOL_EXPORT PeriodicTable {
       }
     }
-    POSTCONDITION(anum > -1, "Element '" + elementSymbol + "' not found");
+    POSTCONDITION_NOLOG(anum > -1, "Element '" + elementSymbol + "' not found");
```
I an early pass, the routine to get the routine ParseV3000AtomSymbol was called and would fail for macroatoms.  The failure was caught, and the label was detected and saved.   But the error message was printed in python that the element name was not known.   This was changed so that if a macro atom is expected, that routine is not called.

Thus POSTCONDITION_NOLOG is not needed, and it has been removed.

tadhurst-cdd commented on Code/RDGeneral/Invariant.h:
```diff
@@ -122,6 +122,13 @@ RDKIT_RDGENERAL_EXPORT std::ostream &operator<<(std::ostream &s,
     throw inv;                                                              \
   }
 
+#define POSTCONDITION_NOLOG(expr, mess)                                     \
```
POSTCONDITION_NOLOG has been removed

tadhurst-cdd commented on Code/GraphMol/ScsrMol.h:
```diff
@@ -0,0 +1,62 @@
+
+class RDKIT_GRAPHMOL_EXPORT SCSRMol {
+ private:
+  std::unique_ptr<ROMol> mol;
```
OK I changed it

tadhurst-cdd commented on Code/GraphMol/ScsrMol.h:
```diff
@@ -0,0 +1,62 @@
+class RDKIT_GRAPHMOL_EXPORT SCSRMol {
+ private:
+  std::unique_ptr<ROMol> mol;
+  std::vector<std::unique_ptr<ROMol>> templates;
```
changed

tadhurst-cdd commented on Code/GraphMol/ScsrMol.h:
```diff
@@ -0,0 +1,62 @@
+  std::vector<std::unique_ptr<ROMol>> templates;
+
+ public:
+  SCSRMol();
```
I honestly did not think about it, but I am not sure it needs to be serialized.

tadhurst-cdd commented on Code/GraphMol/ScsrMol.h:
```diff
@@ -0,0 +1,62 @@
+
+  virtual ~SCSRMol() {}
+
+  void addTemplate(ROMol *templateMol) {
```
OK

tadhurst-cdd commented on Code/GraphMol/ScsrMol.h:
```diff
@@ -0,0 +1,62 @@
+  ROMol *getMol() { return mol.get(); }
+};
+
+typedef boost::shared_ptr<SCSRMol> SCSRMOL_SPTR;
```
I guess it could be, but I found no references to it.  The SCSRMol is an intermediate  entity and it does not seem like it would be shared among different routines.    I have removed it, and its use in .../Code/GraphMol/Wrap/Mol.cpp as the definition of the WRAP SCSRMol, and all test pass.

d-b-w reviewed (commented):
Is the SCSRMol only used as a stop along the route between SCSR files and atomistic mols? Or do you have other internal use of the SCSRMol? Basically, I feel like the rationale and scope of this change would be easier to understand and test if it did not export SCSRMol to Python, and instead SCSRMol were strictly an internal implementation detail.

tadhurst-cdd commented:
> Is the SCSRMol only used as a stop along the route between SCSR files and atomistic mols? Or do you have other internal use of the SCSRMol? Basically, I feel like the rationale and scope of this change would be easier to understand and test if it did not export SCSRMol to Python, and instead SCSRMol were strictly an internal implementation detail.

We do expect that the SCSRMol will be used for sequence-based searching and extraction of the natural anaolog sequence, and more.

tadhurst-cdd commented on Code/GraphMol/GraphMol.h:
```diff
@@ -18,5 +18,6 @@
 
 #include "ROMol.h"
 #include "RWMol.h"
+#include "ScsrMol.h"
```
OK with me - I removed it

tadhurst-cdd commented on Code/GraphMol/FileParsers/testMacromols.cpp:
```diff
@@ -0,0 +1,242 @@
+//
```
I created a catch test, and deleted this one

tadhurst-cdd commented on CMakeLists.txt:
```diff
@@ -3,7 +3,7 @@ cmake_minimum_required(VERSION 3.18)
 
 IF(NOT CMAKE_BUILD_TYPE)
-  SET(CMAKE_BUILD_TYPE Release CACHE STRING
+  SET(CMAKE_BUILD_TYPE Debug CACHE STRING
```
Good catch - thanks.

greglandrum commented:
@tadhurst-cdd, the PDF describing the SCSR support is really helpful, thanks for doing it.
I believe that it will also be useful in the future for people who want to understand the RDKit SCSR support, so it should be in the Docs folder of the repo.
Since PDF isn't the ideal primary doc format: could you do either a markdown (probably easier) or rst version of that document and add it to `Docs/Book`?

greglandrum commented:
> > Is the SCSRMol only used as a stop along the route between SCSR files and atomistic mols? Or do you have other internal use of the SCSRMol? Basically, I feel like the rationale and scope of this change would be easier to understand and test if it did not export SCSRMol to Python, and instead SCSRMol were strictly an internal implementation detail.
> 
> We do expect that the SCSRMol will be used for sequence-based searching and extraction of the natural anaolog sequence, and more.

Wanting to have a monomer-based form of the molecule makes sense, and I agree that it would be useful. That would then need to be fully functional and support things like serialization. I know that you've talked to with @d-b-w and @rachelnwalker about the `MonomerMol` proposal (#8218) and I would like to figure out how we can integrate the two ideas without making the whole thing take a year.

A proposal to start a discussion (we can also schedule a call for this if you think it makes sense): Since these are both large-ish PRs, maybe it would make sense to go stepwise here:
1. Get an initial SCSR implementation in with the less-than-fully-fleshed out SCSRMol as an intermediate data structure that's not exposed to Python (and with a note in the C++ docs that it should not be used in client code)
2. Get the initial MonomerMol implementation in
3. Update the SCSR code to use the MonomerMol
4. (in parallel?) figure out how to adapt the rest of the RDKit's "sequence" parsing/writing code to use MonomerMols.

This would get both the SCSR and MonomerMol support out there for people to try, would reduce duplicated effort, and would hopefully get us something that's more easily supported long term.

tadhurst-cdd commented:
> @tadhurst-cdd, the PDF describing the SCSR support is really helpful, thanks for doing it. I believe that it will also be useful in the future for people who want to understand the RDKit SCSR support, so it should be in the Docs folder of the repo. Since PDF isn't the ideal primary doc format: could you do either a markdown (probably easier) or rst version of that document and add it to `Docs/Book`?

I added a section for SCSR treatment to RDKit_Book.rst like I did for Atropisomers.

tadhurst-cdd commented:
As suggested by @greglandrum ("Get an initial SCSR implementation in with the less-than-fully-fleshed out SCSRMol as an intermediate data structure that's not exposed to Python")
SCSRMol is now not exposed in python and changed the calls in Python to go directly from an SCSR file or text block to a full-atomistic mol.

greglandrum commented:
> As suggested by @greglandrum ("Get an initial SCSR implementation in with the less-than-fully-fleshed out SCSRMol as an intermediate data structure that's not exposed to Python") SCSRMol is now not exposed in python and changed the calls in Python to go directly from an SCSR file or text block to a full-atomistic mol.

Thanks @tadhurst-cdd.
Would you also be agreeable to replacing the functions in `FileParsers.h` that return `SCSRMol` objects with something that returns RWMol instead?
This would be:
- SCSRMolFromSCSRDataStream -> MolFromSCSRDataStream
- SCSRMolFromSCSRBlock -> MolFromSCSRBlock
- SCSRMolFromSCSRFile -> MolFromSCSRFile

And then MolFromSCSRMol could be removed and SCSRMol.h could come out of the installed headers

This would get us something that we could include in the 2025.03.1 release to allow people to convert SCSR to molecules (probably the major use case) while leaving us the flexibility to decide for a future release what we want to use as a monomer-based representation. Once a class is exposed in the public API it is tricky to remove it later.

If you agree that this isn't going to cause problems, I'm happy to do the PR with the changes.

tadhurst-cdd commented:
I can do that.   Later today.

-tad

 

Dr. Tad Hurst

Collaborative Drug Discovery, Inc.

 ***@***.***> ***@***.***

314-610-4313

 

From: Greg Landrum ***@***.***> 
Sent: Wednesday, March 5, 2025 7:14 AM
To: rdkit/rdkit ***@***.***>
Cc: tadhurst-cdd ***@***.***>; Mention ***@***.***>
Subject: Re: [rdkit/rdkit] SCSR capability in RDKit (PR #8147)

 

As suggested by @greglandrum <https://github.com/greglandrum>  ("Get an initial SCSR implementation in with the less-than-fully-fleshed out SCSRMol as an intermediate data structure that's not exposed to Python") SCSRMol is now not exposed in python and changed the calls in Python to go directly from an SCSR file or text block to a full-atomistic mol.

Thanks @tadhurst-cdd <https://github.com/tadhurst-cdd> .
Would you also be agreeable to replacing the functions in FileParsers.h that return SCSRMol objects with something that returns RWMol instead?
This would be:

*	SCSRMolFromSCSRDataStream -> MolFromSCSRDataStream
*	SCSRMolFromSCSRBlock -> MolFromSCSRBlock
*	SCSRMolFromSCSRFile -> MolFromSCSRFile

And then MolFromSCSRMol could be removed and SCSRMol.h could come out of the installed headers

This would get us something that we could include in the 2025.03.1 release to allow people to convert SCSR to molecules (probably the major use case) while leaving us the flexibility to decide for a future release what we want to use as a monomer-based representation. Once a class is exposed in the public API it is tricky to remove it later.

If you agree that this isn't going to cause problems, I'm happy to do the PR with the changes.

—
Reply to this email directly, view it on GitHub <https://github.com/rdkit/rdkit/pull/8147#issuecomment-2700897770> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/A22KQE6BE7NB3LW6JOKELQD2S32C5AVCNFSM6AAAAABUPBPCX6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBQHA4TONZXGA> .
You are receiving this because you were mentioned.  <https://github.com/notifications/beacon/A22KQEY2UO2INRUFMQ4FZFL2S32C5A5CNFSM6AAAAABUPBPCX6WGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTVA7RW6U.gif> Message ID: ***@***.*** ***@***.***> >

  <https://avatars.githubusercontent.com/u/540511?s=20&v=4> greglandrum left a comment (rdkit/rdkit#8147) <https://github.com/rdkit/rdkit/pull/8147#issuecomment-2700897770> 

As suggested by @greglandrum <https://github.com/greglandrum>  ("Get an initial SCSR implementation in with the less-than-fully-fleshed out SCSRMol as an intermediate data structure that's not exposed to Python") SCSRMol is now not exposed in python and changed the calls in Python to go directly from an SCSR file or text block to a full-atomistic mol.

Thanks @tadhurst-cdd <https://github.com/tadhurst-cdd> .
Would you also be agreeable to replacing the functions in FileParsers.h that return SCSRMol objects with something that returns RWMol instead?
This would be:

*	SCSRMolFromSCSRDataStream -> MolFromSCSRDataStream
*	SCSRMolFromSCSRBlock -> MolFromSCSRBlock
*	SCSRMolFromSCSRFile -> MolFromSCSRFile

And then MolFromSCSRMol could be removed and SCSRMol.h could come out of the installed headers

This would get us something that we could include in the 2025.03.1 release to allow people to convert SCSR to molecules (probably the major use case) while leaving us the flexibility to decide for a future release what we want to use as a monomer-based representation. Once a class is exposed in the public API it is tricky to remove it later.

If you agree that this isn't going to cause problems, I'm happy to do the PR with the changes.

—
Reply to this email directly, view it on GitHub <https://github.com/rdkit/rdkit/pull/8147#issuecomment-2700897770> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/A22KQE6BE7NB3LW6JOKELQD2S32C5AVCNFSM6AAAAABUPBPCX6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBQHA4TONZXGA> .
You are receiving this because you were mentioned.  <https://github.com/notifications/beacon/A22KQEY2UO2INRUFMQ4FZFL2S32C5A5CNFSM6AAAAABUPBPCX6WGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTVA7RW6U.gif> Message ID: ***@***.*** ***@***.***> >

tadhurst-cdd commented on Code/GraphMol/ScsrMol.h:
```diff
@@ -0,0 +1,62 @@
+  SCSRMol(SCSRMol &&other) noexcept;
+  SCSRMol &operator=(SCSRMol &&other) noexcept;
+
+  virtual ~SCSRMol() {}
```
I was taking a stab at fixingthe swig error in the CI - it was complaining about calling a deleted function (~SCSRMol) when a unique_ptr was used.  This did not work.  I dont think this needs to be here at all AFAIK

tadhurst-cdd commented:
I changed the new test for macromols to NOT include comparison of full output files.    I did add a check for the number of sgroups created for each test case

greglandrum reviewed (changes requested):
rather than waiting until I'm all the way through, I figure it's worth submitting a review on what I've already read

tadhurst-cdd commented:
> 
> And then MolFromSCSRMol could be removed and SCSRMol.h could come out of the installed headers
> 
SCSRMol is (for ) now only visible inside the SCSRMolFIleParse.cpp

tadhurst-cdd commented on Code/GraphMol/FileParsers/MolFileParser.cpp:
```diff
@@ -1380,10 +1380,13 @@ void setRGPProps(const std::string_view symb, Atom *res) {
 
-void lookupAtomicNumber(Atom *res, const std::string &symb,
-                        bool strictParsing) {
+void lookupAtomicNumber(Atom *res, std::string &symb, bool strictParsing) {
```
I have no idea, and I don't remember making such a change.  I have changed it back.

tadhurst-cdd commented on Code/GraphMol/FileParsers/MolFileParser.cpp:
```diff
@@ -2381,7 +2378,45 @@ void ParseV3000AtomProps(RWMol *mol, Atom *&atom, typename T::iterator &token,
       }
     } else if (prop == "ATTCHORD") {
-      if (val != "0") {
+      // there are two kinds of ATTAHORD
+      // one is for template instances and look like this: ATTCHORD=(4 1 Al 3
+      // Br)
```
Yes

tadhurst-cdd commented on Code/GraphMol/FileParsers/MolFileParser.cpp:
```diff
@@ -2523,7 +2558,35 @@ void ParseV3000AtomBlock(std::istream *inStream, unsigned int &line,
+
+    auto isMacroAtom = false;
+    if (expectMacroAtoms) {
+      std::vector<std::string_view>::iterator lookAheadToken = token + 1;
```
OK done

tadhurst-cdd commented on Code/GraphMol/FileParsers/MolFileParser.cpp:
```diff
@@ -3195,37 +3260,87 @@ bool ParseV3000CTAB(std::istream *inStream, unsigned int &line, RWMol *mol,
+    if (tempStr.length() >= 12 && tempStr.substr(0, 12) == "BEGIN SGROUP") {
+      if (sgroupFound) {
+        std::ostringstream errout;
+        errout << "BEGIN SGROUP found more than once on line " << line;
```
SGroups and Collections could occur in any order (but not twice), so the code was changed NOT to expect them in a particular order.   There is a test file with more than one occurrence of SGroups.  II think I created it just to test this check.

tadhurst-cdd commented on Code/GraphMol/FileParsers/SCSRMolFileParser.cpp:
```diff
@@ -0,0 +1,1455 @@
+
+  res->getMol()->setProp(common_properties::_Name, tempStr);
+
+  // info
```
I fixed this to use one sourse with a flag to indicate that it is parsing an SCSR main block.

tadhurst-cdd commented on Code/GraphMol/FileParsers/CMakeLists.txt:
```diff
@@ -5,7 +5,8 @@ endif()
 
 rdkit_library(FileParsers
```
OK done

tadhurst-cdd commented on Code/GraphMol/Aromaticity.cpp:
```diff
@@ -665,6 +665,35 @@ int countAtomElec(const Atom *at) {
   return res;
 }
 
+static bool molHasHydrogenBond(const ROMol &mol) {
```
Done

tadhurst-cdd commented on Code/GraphMol/SCSRMol.h:
```diff
@@ -0,0 +1,63 @@
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+/*! \file RWMol.h
+
+  \brief Defines the editable molecule class \c RWMol
+
+*/
+
```
This file no longer exists.

tadhurst-cdd commented on Code/GraphMol/SCSRMol.cpp:
```diff
@@ -0,0 +1,44 @@
+//
+//  Copyright (C) 2003-2024 Tad Hurst and other RDKit contributors
```
this file no longer exists - but I did change this in Code/GraphMol/FileParsers/SCSRMolFileParser.cpp

tadhurst-cdd commented on Code/GraphMol/SCSRMol.h:
```diff
@@ -0,0 +1,63 @@
+  SCSRMol &operator=(SCSRMol &&other) noexcept;
+
+  SCSRMol &operator=(const SCSRMol &) = delete;  // disable assignment
+  virtual void destroySCSR();
+  virtual ~SCSRMol() { destroySCSR(); }
```
I was trying to get the CI to complete when I checked things in.  THis file no longer exists, and I have made the destructor non-virtual.  I also disabled other non-used constructors.

greglandrum commented on Docs/Book/RDKit_Book.rst:
```diff
@@ -2701,6 +2701,206 @@ path) enumeration algorithm used in the RDKit fingerprint. After a subgraph has
+“CTfile Formats - Dassault Systèmes”.  That document does not provide any real detail for that format, and contains one example file.
+
+In addition, Biovia Draw supports reading and writing this format.  As much as possible, the RDKit support allows the functionality
+supported by Biovia Draw.  One exception is the RDkit treatment of hydrogen bonds in SCSR files/blocks (vide infra).
```
```suggestion
supported by Biovia Draw.  One exception is the RDKit treatment of hydrogen bonds in SCSR files/blocks (vide infra).
```

tadhurst-cdd replied:
ok

greglandrum commented on Docs/Book/RDKit_Book.rst:
```diff
@@ -2701,6 +2701,206 @@ path) enumeration algorithm used in the RDKit fingerprint. After a subgraph has
 been generated, it is used to set multiple bits based on different atom and bond
 type definitions.
 
+Self-Contained Structure Representations (SCSR) for Macromolecules
```
Thanks for writing this. It's really useful

greglandrum commented on Docs/Book/RDKit_Book.rst:
```diff
@@ -2701,6 +2701,206 @@ path) enumeration algorithm used in the RDKit fingerprint. After a subgraph has
+An SCSR file contains a mol with a CTAB in v3000 format.   That CTAB can contain elemental atoms and macro atoms corresponding to amino acids (AA), RNA, and DNA elements.  
+The RNA and DNA elements are represented by three parts – a phosphate group, a sugar, and a base.  
+
+Each atom line in the CAT can refer to an elemental atom or a macro atom.  Macro atom lines have a text description of the macro name and must have a CLASS and an
```
```suggestion
Each atom line in the CTAB can refer to an elemental atom or a macro atom.  Macro atom lines have a text description of the macro name and must have a CLASS and an
```

tadhurst-cdd replied:
ok

greglandrum commented on Docs/Book/RDKit_Book.rst:
```diff
@@ -2701,6 +2701,206 @@ path) enumeration algorithm used in the RDKit fingerprint. After a subgraph has
+Each atom line in the CAT can refer to an elemental atom or a macro atom.  Macro atom lines have a text description of the macro name and must have a CLASS and an
+ATTCHORD attributed.  They can also have an optional SEQID attribute.  
+
+According to the Biovia doc, the CLASS attribute must have a value that is one of:  AA, dAA, ScsrSupportInRDKitDNA, RNA, SUGAR, BASE, PHOSPHATE, LINKER, CHEM, LGR,P MODAA, MODdAA, MODDNA, MODRNA, XLINKAA, XLINKdAA, XLINKDNA, XLINKRNA.   
```
```suggestion
According to the Biovia doc, the CLASS attribute must have a value that is one of:  AA, dAA, DNA, RNA, SUGAR, BASE, PHOSPHATE, LINKER, CHEM, LGRP,  MODAA, MODdAA, MODDNA, MODRNA, XLINKAA, XLINKdAA, XLINKDNA, XLINKRNA.   
```

tadhurst-cdd replied:
OK

greglandrum commented on Docs/Book/RDKit_Book.rst:
```diff
@@ -2701,6 +2701,206 @@ path) enumeration algorithm used in the RDKit fingerprint. After a subgraph has
+ATTCHORD attributed.  They can also have an optional SEQID attribute.  
+
+According to the Biovia doc, the CLASS attribute must have a value that is one of:  AA, dAA, ScsrSupportInRDKitDNA, RNA, SUGAR, BASE, PHOSPHATE, LINKER, CHEM, LGR,P MODAA, MODdAA, MODDNA, MODRNA, XLINKAA, XLINKdAA, XLINKDNA, XLINKRNA.   
+This implementation in RDKit does NOT enforce the allowed list.
```
I think it does enforce the allowed list.

tadhurst-cdd replied:
SO I took a look.  The class in an atom definition is NOT checked.  There is an example in Code/GraphMol/FileParsers/file_parsers_catch.cpp that has a class of "foo".

The class in an SGROUP is checked.   In SCSR block parsing, the atom line will have a class which is not checked against the allowed list, but this must also appear in an SGROUP and that is checked, so the SCSR treatment DOES enforce the allowed list.  

Thanks for catching this.  I have changed the doc to reflect this.

greglandrum commented on Docs/Book/RDKit_Book.rst:
```diff
@@ -2701,6 +2701,206 @@ path) enumeration algorithm used in the RDKit fingerprint. After a subgraph has
+BEGIN TEMPLATES line and an END TEMPLATES line. Each template starts with a TEMPLATE line that indicates the template number (1 to n), the template Class and 
+Name, and the NATREPLACE attribute.
+
+The Class and Name consists of three (or more) parts separated by forward slashes.  The first part is the CLASS,  and must match the CLASS attribute in 
```
```suggestion
The Class and Name consists of three (or more) parts separated by forward slashes.  The first part is the CLASS, and must match the CLASS attribute in 
```

tadhurst-cdd replied:
OK

greglandrum commented on Docs/Book/RDKit_Book.rst:
```diff
@@ -2701,6 +2701,206 @@ path) enumeration algorithm used in the RDKit fingerprint. After a subgraph has
+CTAB, and a string that indicates the attachment point for this macro atom.   The attachment point string is always two characters.  
+The first indicates the order, and is a capital letter starting at A.  The second char is one of “l” (left), “r” (right), “x” (cross connections – e.g. for cysteine).  
+In this implementation, the second character can also be “h” for hydrogen bond.   
+Thus, the attach connections are almost always in the list: “Al”, “Br”, “Cx” or “Ch”.   For example:
+
+    # ATTCHORD=(6 15 Br 13 Al 21 Cx)
```
It's not clear to me why this starts with a `#`.
In order to get these to render as code samples, which I think is what you want, you need to end the previous line with `::` (yes, RST is irritating).
So this becomes:
```suggestion
Thus, the attach connections are almost always in the list: “Al”, “Br”, “Cx” or “Ch”.   For example::

    ATTCHORD=(6 15 Br 13 Al 21 Cx)
```

Please update all the other examples too.

tadhurst-cdd replied:
OK

greglandrum commented on Docs/Book/RDKit_Book.rst:
```diff
@@ -2701,6 +2701,206 @@ path) enumeration algorithm used in the RDKit fingerprint. After a subgraph has
+  * SCSRMolFromScsrBlock  (text block to SCSRMol)
+  * SCSRMolFromScsrFile  (file to SCSRMol)
+  
+Full Example of SCST files:
+===========================
+
+Here are two examples of SCST files:
```
```suggestion
Full Example of SCSR files:
===========================

Here are two examples of SCSR files:
```

tadhurst-cdd replied:
OK

greglandrum commented on Docs/Book/RDKit_Book.rst:
```diff
@@ -2701,6 +2701,206 @@ path) enumeration algorithm used in the RDKit fingerprint. After a subgraph has
+
+Here are two examples of SCST files:
+
+`DnaTest3 </Docs/Book/data/DnaTest3.mol>`_.
+
+`CrossLink.mol </Docs/Book/data/CrossLink.mol>`_.
```
These links do not work in the final documentation.

You either need to figure out how to make Sphinx copy the files and what the format of the links should be or include links to where the files will be in Github, like this:
(i.e. ``)
```suggestion
`DnaTest3.mol <https://github.com/rdkit/rdkit/blob/master/Docs/Book/data/DnaTest3.mol>`_

`CrossLink.mol <https://github.com/rdkit/rdkit/blob/master/Docs/Book/data/CrossLink.mol>`_
```

tadhurst-cdd replied:
I made these changes as suggested.  That link does not work for me either because those files are not in master yet

greglandrum commented on Docs/Book/RDKit_Book.rst:
```diff
@@ -2701,6 +2701,206 @@ path) enumeration algorithm used in the RDKit fingerprint. After a subgraph has
+
+    # M  V30 1 SUP 1 ATOMS=(1 11) XBONDS=(1 11) LABEL=H CLASS=LGRP
+
+Parsing SCSR files and text blocks
+==================================
+
+The following calls parse the SCSR mol or block produces an SCSRMol:
+
+  * SCSRMolFromScsrDataStream  (stream to SCSRMol)
+  * SCSRMolFromScsrBlock  (text block to SCSRMol)
+  * SCSRMolFromScsrFile  (file to SCSRMol)
```
This block documents functionality that's not in the public API, please remove

greglandrum commented on Docs/Book/RDKit_Book.rst:
```diff
@@ -2701,6 +2701,206 @@ path) enumeration algorithm used in the RDKit fingerprint. After a subgraph has
+`CrossLink.mol </Docs/Book/data/CrossLink.mol>`_.
+
+  
+Expansion to Full Atomistic Form
+===============================
```
This block should be adapted to document the functions which are actually in the public API.

tadhurst-cdd replied:
reworked the doc to reflect the actual calls that are available

greglandrum replied:
> reworked the doc to reflect the actual calls that are available

Great, please let me know when you have pushed the changes and I will take a look.

greglandrum commented on Docs/Book/RDKit_Book.rst:
```diff
@@ -2701,6 +2701,206 @@ path) enumeration algorithm used in the RDKit fingerprint. After a subgraph has
+    ScsrTemplateNamesUseFirstName, or ScsrTemplateNamesUseLastName.   If ScsrTemplateNamesAsEntered is specified, the name as referenced in the main 
+    SCSR CTAB will be used.  
+
+    * The ”includeLeavingGroups” parameter control whether leaving groups that are not replaced in the main CTAB are included in the resulting atomistic file.   
```
```suggestion
    * The ”includeLeavingGroups” parameter controls whether leaving groups that are not replaced in the main CTAB are included in the resulting atomistic file.   
```

tadhurst-cdd replied:
ok

tadhurst-cdd commented on .azure-pipelines/linux_build.yml:
done

tadhurst-cdd commented on Code/GraphMol/Bond.h:
```diff
@@ -386,12 +386,16 @@ class RDKIT_GRAPHMOL_EXPORT Bond : public RDProps {
 
-inline bool isDative(const Bond &bond) {
-  auto bt = bond.getBondType();
+inline bool isDative(const Bond::BondType &bt) {
```
ok - remove the reference marker

tadhurst-cdd commented on Code/GraphMol/Wrap/enhanced_stereo_test.py:
Ha!  this is a variant of your code to test the effeciency of enhanced stereo, and I used it for my testing.  It should not be there.  I have removed this file

tadhurst-cdd commented:
I have pushed the latest changes to PR:

https://github.com/rdkit/rdkit/pull/8147

 

thanks!

 

tad

 

Dr. Tad Hurst

Collaborative Drug Discovery, Inc.

 ***@***.***> ***@***.***

314-610-4313

 

From: Greg Landrum ***@***.***> 
Sent: Saturday, May 10, 2025 10:20 PM
To: rdkit/rdkit ***@***.***>
Cc: tadhurst-cdd ***@***.***>; Mention ***@***.***>
Subject: Re: [rdkit/rdkit] SCSR capability in RDKit (PR #8147)

 

@greglandrum commented on this pull request.

  _____  

In Docs/Book/RDKit_Book.rst <https://github.com/rdkit/rdkit/pull/8147#discussion_r2083384601> :

> +Expansion to Full Atomistic Form
+===============================

reworked the doc to reflect the actual calls that are available

Great, please let me know when you have pushed the changes and I will take a look.

—
Reply to this email directly, view it on GitHub <https://github.com/rdkit/rdkit/pull/8147#discussion_r2083384601> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/A22KQE23NK2XUOGPPH23P2D2526U5AVCNFSM6AAAAABUPBPCX6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDQMZRGIYTGMRSGQ> .
You are receiving this because you were mentioned.  <https://github.com/notifications/beacon/A22KQE5CN77Z2AN2SFLE5LD2526U5A5CNFSM6AAAAABUPBPCX6WGG33NNVSW45C7OR4XAZNRKB2WY3CSMVYXKZLTORJGK5TJMV32UY3PNVWWK3TUL5UWJTVIYDRKQ.gif> Message ID: ***@***.*** ***@***.***> >

greglandrum reviewed (approved):
LGTM
:fireworks:

bp-kelley commented:
Great job everyone, we’re looking forward to this.

greglandrum commented:
Yeah, thanks for all of the work on this @tadhurst-cdd!
